torch/_inductor/compile_fx_async.py

```diff
@@ -179,3 +179,183 @@ def callback(pickled_output: _WireProtocolPickledOutput) -> OutputCode:
             return output.graph
 
         return _AsyncOutputCode(eager_output_code, f, callback)
+
+
+# _ProgressiveOutputCode handles running a fast compile first, then hot-swapping
+# to a more optimized version when the expensive compile finishes.
+@final
+class _ProgressiveOutputCode(OutputCode):
+    _fast_output_code: Optional[OutputCode]
+    _optimized_output_code: Optional[OutputCode]
+    _progression_futures: list[Future[_WireProtocolPickledOutput]]
+    _callback: Callable[[_WireProtocolPickledOutput], OutputCode]
+    _post_compile_data: Optional[_PostCompileData] = None
+    _boxed_call: bool
+    _current_progression_index: int
+
+    def __init__(
+        self,
+        # Fast compile that runs faster than the progressive compiles
+        fast_output_code: OutputCode,
+        # Futures for the progressive optimized compiles
+        progression_futures: list[Future[_WireProtocolPickledOutput]],
```

Comment:
It seems like some users could want this, where we kick off N optimized compiles in parallel, but I could also envision some where they want N optimized compiles kicked off successively only once the previous ones finish. What should the API look like? Does it take `Future[...] | Callable[..., Future[...]]`?

Reply:
This will presumably all be remote, so it seems unlikely to be a common request. I could imagine it coming up if users are concerned about capacity, but to avoid premature optimization, let's hold off and add it if and when users ask for it?

These are all private APIs so we don't need to worry about BC and can always change it later.